examples/repeater/second_gen_with_repetition_code.py

```diff
@@ -0,0 +1,246 @@
+        return self.physical[index]
+
+    def H(self):  # this maps ⌈|0>⌋ to ⌈|+>⌋
+        self.physical[0].H()
+        self.physical[0].cnot(self.physical[1])
+        self.physical[0].cnot(self.physical[2])
+
+
+# This circuit is from https://arxiv.org/pdf/quant-ph/0002039.pdf, page 9. It
+# lets two peers perform a remote CNOT using a single EPR pair.
+@dataclass()
+class RemoteCNOT:
+    """
+    RemoteCNOT teleports qubits, |alpha> and |beta>, through an EPR pair
+    composed of |red> and |blue>. The result of the protocol is that
+        |red>  = |alpha>
+        |blue> = |beta ⊕ alpha>
+    """
+    left: Host
+    right: Host
+
+    def left_protocol(self, alpha: Qubit, red: Qubit):
+        red.cnot(alpha)
+        x = alpha.measure()
+        if x == 1:
+            red.X()
+        self.left.send_classical(self.right.host_id, str(x), await_ack=True)
+
+        z = self.left.get_next_classical(self.right.host_id, wait=-1).content
+        if z == '1':
+            red.Z()
+
+    def right_protocol(self, blue: Qubit, beta: Qubit):
+        beta.cnot(blue)
+        beta.H()
+
+        x = self.right.get_next_classical(self.left.host_id, wait=-1).content
+        if x == '1':
+            blue.X()
+
+        z = beta.measure()
+        self.right.send_classical(self.left.host_id, str(z), await_ack=True)
+        if z == 1:
+            blue.Z()
+
+
+@dataclass()
+class EncodedGenerationProtocol:
+    """
+    EncodedGenerationProtocol establishes an encoded Φ^+ between left and right
+    hosts. This is done as follows.
+
+    1. locally prepare encoded states ⌈|+>⌋ and ⌈|0>⌋
+
+    For each each physical qubit:
+
+    2. left creates an EPR pair
+    3. left distributes half of the EPR pair to right
+    4. peers use the EPR pair to perform a transverse teleportation-based CNOT
+    """
+    left: Host
+    right: Host
+    remote_cnot: RemoteCNOT
+
+    def __init__(self, left, right):
+        self.left = left
+        self.right = right
+        self.remote_cnot = RemoteCNOT(left, right)
+
+    def left_protocol(self, left: Host, n: int):
+        logical = RepetitionCodedQubit(self.left)
+        logical.H()
+
+        for k, physical in enumerate(logical):
+            epr = send_epr(self.left, self.right)
+            self.remote_cnot.left_protocol(physical, epr)
+            self.left.add_qubit(self.left.host_id, epr, f"{n}>{k}")
+
+    def right_protocol(self, right: Host, n: int):
+        # prepare an encoded |0>
+        logical = RepetitionCodedQubit(right)
+
+        for k, physical in enumerate(logical):
+            epr = self.right.get_qubit(self.left.host_id, wait=-1)
+            self.remote_cnot.right_protocol(epr, physical)
+            right.add_qubit(right.host_id, epr, f"{n}<{k}")
+
+    def get_logical_qubit(self, host, n):
+        return (host.get_qubit_by_id(f"{n}>{k}") for k in range(3))
+
+
+def encoded_connection(host: Host, left: Host, right: Host, n: int):
+    ms = []
+    for k in range(3):
```

Comment:
Can the 3 be generalized? It's a bit of a "magic number"

Reply:
Done!

I'd be nice to have an abstraction for getting/storing logical qubits. But I reckon that's outside the scope of this example.